src/types/acls.rs

```diff
@@ -0,0 +1,287 @@
+pub use imap_proto::types::AclRight;
+
+/// enum used for set_acl to specify how the ACL is to be modified.
+pub enum AclModifyMode {
+    /// Replace all ACLs on the identifier for the mailbox
+    Replace,
+    /// Add the given ACLs to the identifier for the mailbox
+    Add,
+    /// Remove the given ACLs from the identifier for the mailbox
+    Remove,
+}
+
+/// Helpful wrapper around the ACL rights vector
+#[derive(Debug, Eq, PartialEq)]
+pub struct AclRightList {
+    pub(crate) data: HashSet<AclRight>,
+}
+
+impl AclRightList {
+    /// Returns if the AclRightList has the provided ACL (either as a char or an AclRight enum)
+    pub fn has_right<T: Into<AclRight>>(&self, right: T) -> bool {
+        self.data.contains(&right.into())
+    }
+}
+
+impl ToString for AclRightList {
+    fn to_string(&self) -> String {
+        let mut v: Vec<char> = self.data.iter().map(|c| char::from(*c)).collect();
+
+        v.sort_unstable();
+
+        v.into_iter().collect()
+    }
+}
+
+impl From<HashSet<AclRight>> for AclRightList {
+    fn from(hash: HashSet<AclRight>) -> Self {
+        Self { data: hash }
+    }
+}
+
+impl From<Vec<AclRight>> for AclRightList {
+    fn from(vec: Vec<AclRight>) -> Self {
+        AclRightList {
+            data: vec.into_iter().collect(),
+        }
+    }
+}
+
+impl From<&str> for AclRightList {
+    fn from(i: &str) -> Self {
+        i.chars()
+            .into_iter()
+            .map(|c| c.into())
+            .collect::<HashSet<AclRight>>()
+            .into()
+    }
+}
+
+/// From [section 3.6 of RFC 4313](https://datatracker.ietf.org/doc/html/rfc4314#section-3.6).
+///
+/// The ACL response from the getacl IMAP command
+#[derive(Debug, Eq, PartialEq)]
+pub struct Acl {
+    /// the mailbox these rights list are for
+    pub mailbox: String,
+    /// The list of identifier/rights pairs for the mailbox
+    pub acls: Vec<AclEntry>,
+}
+
+impl Acl {
+    /// Parse the given input into a [`ACL`] response.
+    pub fn parse(
+        lines: Vec<u8>,
+        unsolicited: &mut mpsc::Sender<UnsolicitedResponse>,
+    ) -> Result<Self, Error> {
+        let mut lines: &[u8] = &lines;
+        let mut acl = None;
+
+        while !lines.is_empty() {
+            match imap_proto::parser::parse_response(lines) {
+                Ok((rest, Response::Acl(a))) => {
+                    lines = rest;
+                    acl = Some(a.into());
+                },
+                Ok((rest, data)) => {
+                    lines = rest;
+                    if let Some(resp) = try_handle_unilateral(data, unsolicited) {
+                        return Err(resp.into());
+                    }
+                }
+                _ => {
+                    return Err(Error::Parse(ParseError::Invalid(lines.to_vec())));
+                }
+            }
+        }
+
+        acl.ok_or_else(|| Error::Parse(ParseError::Invalid(lines.to_vec())))
+    }
+}
+
+impl Into<Acl> for imap_proto::types::Acl<'_> {
```

Comment:
We should avoid this `impl` and make it an inherent method instead since it ends up being another public part of the API tied to `imap_proto`.

Reply:
@jonhoo is something like this what you are meaning here? d0a0f0d530

Reply:
Yep, exactly!